python/ray/_private/services.py

```diff
@@ -523,8 +523,13 @@ def get_node_to_connect_for_driver(
     filtered_node_to_connect_infos = []
     for node_info in node_to_connect_infos:
         if (
-            node_ip_address is None or node_info.node_manager_address == node_ip_address
-        ) and (node_name is None or node_info.node_name == node_name):
+            (
+                node_ip_address is None
+                or node_info.node_manager_address == node_ip_address
+            )
+            and (node_name is None or node_info.node_name == node_name)
+            and (temp_dir is None or node_info.temp_dir == temp_dir)
```

Comment:
We filter by `temp_dir` here as it's currently possible for users to specify `temp_dir` to `ray.init()`. In this scenario, if there are multiple ray nodes on the same host, `node_ip_address` and `node_name` may be identical, but we would still want the user to connect to the correct node with their desired `temp_dir`.